### DISCUSSION TITLE
NSPanel: Use HMI elements in Tasmota rules

### DISCUSSION THREAD

**Discussion by atirage** [arendst/Tasmota#17119] [Q&A] [answered]
Hi!
I'm trying to use the HMI elements, i.e. switches to trigger some rules.
For ex, the below one is working nicely,

`Rule1 ON NSPanel#params#switches DO WebQuery http://192.168.1.3:8081 POST {"deviceid": "","data": {"switch": "%value%"}} ENDON`

 but it triggers on any switch as the JSON looks like this:

`{"NSPanel":
	{"ctype":"group",
	 "id":"1",
	 "params":
		{"switch":"on",
		 "switches":[{"switch":"on","outlet":0}]
		}
	}
}`

So, my question is how could one identify the switch too using the capabilities of the rules?
The only idea I have, is with at least two rules, one to identify the outlet, then use a var and trigger the 2nd rule to perform the action.
In this case, how to identify the outlet as this is not working:
`ON NSPanel#?#outlet==0 ...` ?

Does anyone have a better approach?
Thanks!

**Comment by sfromis** [comment, marked as answer]
If you are running on an ESP32 device like the NSPanel (?), you have much better options for JSON handling in the Berry language, which I'd recommend.

I'd not expect your JSON there to be supported by rules, as my expectation is that the very limited "classic rules" support for arrays is only for lowest JSON level, not including further key:value pairs inside an array.

**Comment by atirage** [comment]
I considered that as well, I was just thinking about the execution time constraints of the Berry scripts and I thought the rules would be more efficient since currently the JSON is not parsed in the driver code, it is just passed as a string to Tasmota for rule handling and optionally MQTT publishing.
I guess I'll give Berry a try anyway, looks like fun. :)

**Reply by sfromis** [reply]
Do not worry about the negligible execution time of typical Berry code, of course unless you are making time-consuming coding. And the rules handling do need JSON parsing of all payloads, not just the ones you're interested in.

**Comment by atirage** [comment]
I've written my custom Berry code to capture the HMI event and trigger the actions, but now I'm facing the following issue: I have to communicate using REST API and apparently the webclient class has no PUT method, just GET and POST, although in this header it's defined:

> https://github.com/arendst/Tasmota/blob/development/lib/libesp32/Berry-HttpClientLight/src/HttpClientLight.h

Can anyone shed some light on this?

**Reply by barbudor** [reply]
Indeed the PUT is not mapped to Berry

@s-hadinger would you be able to add it ?

**Reply by barbudor** [reply]
I could try otherwise 
Never PRed on Berry 🥶

**Reply by sfromis** [reply]
If PUT is [implemented](https://github.com/arendst/Tasmota/blob/development/lib/libesp32/berry_tasmota/src/be_webclient_lib.c) to mimic REST, maybe also DELETE and PATCH

**Reply by barbudor** [reply]
I'm trying now....
PATCH is already there but not DELETE which is strange as I have already used API with DELETE but never PATCH so far

**Reply by barbudor** [reply]
On going : https://github.com/arendst/Tasmota/pull/17159

If you are in a hurry you can pull my branch: https://github.com/barbudor/Tasmota/tree/berry_wc_put_delete_patch

Very quick and dirty testing testing that the payload to a webserver is correct.

**Reply by s-hadinger** [reply]
Merged

**Reply by atirage** [reply]
Thanks guys for the quick response! I guess there'll be an official release in the near future, so I'll wait for that.

**Reply by sfromis** [reply]
The development precompiled binaries have been updated since the merge, meaning that you can upgrade and run now.

**Reply by atirage** [reply]
Works like a charm! 👍 
The last piece of the puzzle 😎 : if I activate one of the switches, except the one from the 1st index [(NSPanel)](https://blakadder.github.io/nspanel/) , after a certain delay (a few hundred ms) it goes back by itself to off position. As far as I understand the original Berry driver only draws the widgets at screen init, so I suspect this is a firmware behavior of the screen, which would suggest that I have to explicitly send a command to the screen to keep the widget in the wanted state once the triggered command is executed successfully.
Could anyone confirm my assumption?

**Reply by sfromis** [reply]
If you watch the Tasmota console, as it happens, you may get clues about the cause, if you already activated `Weblog 3` for details.

**Reply by atirage** [reply]
I tried your suggestion, but I see only the expected messages:

> 22:19:12.122 NSP: Received Raw = bytes('55AA8669007B226374797065223A2267726F7570222C226964223A2232222C22...')
>22:19:12.139 http://192.168.1.18/things/http---192.168.1.2-things-LEDs/properties/onTV {"onTV":false}
>22:19:12.204 HTP: Connected in 63 ms, stack low mark 3716
>22:19:12.208 http://192.168.1.18/things/http---192.168.1.2-things-LEDs/properties/onBar {"onBar":true}
>22:19:12.264 HTP: Connected in 53 ms, stack low mark 3716
>22:19:12.268 RSL: RESULT = {"NSPanel":{"ctype":"group","id":"2","params":{"switches":[{"switch":"off","outlet":0},"switch":"on","outlet":1}]}}}'

On the display, outlet 1 is the one which goes back to off.

So I think it's safe to say that the Tasmota firmware is not triggering this unexpected toggle.

**Reply by atirage** [reply]
I can confirm my original assumption now, the display does need a confirmation via a so called relation command to keep the widget in the wanted state, if no confirmation comes until the delay expires, it toggles the widget back to its original state.
I'm happy now 😄

**Comment by julianHH** [comment]
Hello @atirage 
would you mind sharing the Berry code you are using, as I am trying to accomplish the same task.
Thank you!

**Reply by atirage** [reply]
Hi,

here are the relevant code snippets:
```
def my_actions(panel, payload)
    import json
    var action = json.load(payload)
    var dummy = 1
    #check if horizontal switch HMI
    if action['NSPanel']['ctype'] == 'group'
        var index = int(action['NSPanel']['id'])
        if index == nil || index <= 0 || index > 4
            return
        end
        index = index - 1
        var outlets = action['NSPanel']['params']['switches']
        var bool_as_str = 'false'
        for i: 0 .. size(outlets) - 1
            #check for a change
            if outlets[i]['switch'] == 'on'
                if horiz_outlets_prev[index][outlets[i]['outlet']] == 1
                    continue
                end
                horiz_outlets_prev[index][outlets[i]['outlet']] = 1
                bool_as_str = 'true'
            else
                if horiz_outlets_prev[index][outlets[i]['outlet']] == 0
                    continue
                end
                horiz_outlets_prev[index][outlets[i]['outlet']] = 0
            end
            #handle each outlet
            var r = -1           
            if index == 0
                if outlets[i]['outlet'] == 0
                    #kitchen
                    r = send_prop_rest('sonoff-diy-adapter-0', 'on', '{"on":' + bool_as_str + '}')
                end
            elif index == 1
                #ambient lights
                if outlets[i]['outlet'] == 0
                    r = send_prop_rest('http---' + thing_ip + '-things-LEDs', 'onTV', '{"onTV":' + bool_as_str + '}')
                elif outlets[i]['outlet'] == 1
                    r = send_prop_rest('http---' + thing_ip + '-things-LEDs', 'onBar', '{"onBar":' + bool_as_str + '}')
                end
            elif index == 2
                if outlets[i]['outlet'] == 0
                    r = send_prop_rest('miLight-adapter-0', 'on', '{"on":' + bool_as_str + '}')
                elif outlets[i]['outlet'] == 1
                    dummy = 0
                elif outlets[i]['outlet'] == 2
                    dummy = 0
                end
            elif index == 3
                if outlets[i]['outlet'] == 0
                    dummy = 0
                elif outlets[i]['outlet'] == 1
                    dummy = 0
                elif outlets[i]['outlet'] == 2
                    dummy = 0
                elif outlets[i]['outlet'] == 3
                    dummy = 0
                end
            else
                dummy = 0
            end
            if r == 200
              #send back OK confirmation to keep display switch on 
              panel.send('{"relation":{"id":"' + action['NSPanel']['id'] + '","params":{"switches":[{"switch":"' + outlets[i]['switch'] + '","outlet":' + str(outlets[i]['outlet']) + '}]}}}')
            elif r != -1
              log('Error sending REST API!')
            end
        end    
    end
 end
```

This function is called here:
```
if msg == bytes('7B226572726F72223A307D') # don't publish {"error":0}
 else
     var jm = string.format("{\"NSPanel\":%s}",msg.asstring())
     my_actions(self, jm)
     tasmota.publish_result(jm, "RESULT")
 end
```
`send_prop_rest` is just a function that prepares and makes a REST API call.
Hope it helps!

**Reply by ip6li** [reply]
How did you initialized horiz_outlets_prev?

**Reply by atirage** [reply]
In the meantime I changed a bit my code, due to a change in house setup, but basically you initialize it to 0, considering which widgets having outlets you are using.  See below:
```
var widget = {
# 1 = toggle switch horizontal
# 2 = toggle switch double horizontal
# 3 = toggle switch triple horizontal
# 4 = toggle switch quad horizontal
# 6 = toggle switch vertical
# 7 = toggle switch double vertical
# 8 = toggle switch triple vertical
# 9 = toggle switch quad vertical
# 33 = RGB light strip
# 52 = CCT bulb
# 69 = RGB+CCT bulb
# leave empty brackets if you don't want a widget there
# ctype scene doesn't have an uiid
# index "name   ", "ctype", uiid | name max 8 characters, rest will be truncated)
# ctype: group, device, scene
  1: ["Kitchen",  "group", 1],
  2: ["Ambient", "group", 2],
  3: ["Hallway",  "group", 1],
  4: ["Living",   "group", 69],
  5: [],
  6: [],
  7: [],
  8: [],
}
# match number of outlets from widget
var outlets_prev = [[0],[0,0],[0]]
```

**Comment by julianHH** [comment]
Thank you very much, @atirage !
